In [1]:
import numpy as np
from sklearn.decomposition import PCA
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

def load_data_from_file(file='preprocessed_data_cropped.npz'):
    data = np.load(file)
    return data['X'], data['y']

# Load the data
X, y = load_data_from_file()

# Flatten the image data
X = X.reshape(X.shape[0], -1)

# Convert one-hot encoded labels back to their original class numbers
y = np.argmax(y, axis=1)

# Split the data into training and testing sets
X_train = X[:39209]
y_train = y[:39209]
X_test = X[39209:]
y_test = y[39209:]


# Define the SVM model with a linear kernel
model = svm.SVC(kernel='linear')
model.fit(X_train, y_train)

# Predict and evaluate accuracy
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("SVM Test Accuracy:", accuracy)


Test Accuracy: 0.876959619952494


 SVM test Accuracy: 0.876959619952494

In [1]:
import numpy as np
from sklearn.model_selection import KFold
from sklearn.svm import SVC
from sklearn import svm
from sklearn.model_selection import KFold
from sklearn.metrics import accuracy_score, f1_score
def load_data_from_file(file='preprocessed_data.npz'):
    data = np.load(file)
    return data['X'], data['y']

# Load the data
X, y = load_data_from_file()

# Reshape the input data into a 2-dimensional array
X = X.reshape(X.shape[0], -1)
# Convert y to a 1D array
y = np.argmax(y, axis=1)
#only use train set to validation
X = X[:39209]
y = y[:39209]
# Define the SVM model with a linear kernel
model = svm.SVC(kernel='linear')

# Define the number of folds for cross-validation
num_folds = 5

# Define the KFold object for splitting the data
kfold = KFold(n_splits=num_folds, shuffle=True, random_state=42)

# Initialize a list to store the accuracy scores for each fold
accuracy_scores = []
f1_scores = []
# Loop through each fold
for fold, (train_index, validation_index) in enumerate(kfold.split(X)):

    # Split the data into training and validationing sets for the current fold
    X_train, X_validation = X[train_index], X[validation_index]
    y_train, y_validation = y[train_index], y[validation_index]

    # Fit the SVM model on the training set
    model.fit(X_train, y_train)

    # Make predictions on the validationing set
    y_pred = model.predict(X_validation)
    # Calculate the accuracy score for the current fold and add it to the list
    accuracy_scores.append(accuracy_score(y_validation, y_pred))
    f1_scores.append(f1_score(y_validation, y_pred, average="weighted"))
    # Print the accuracy score for the current fold
    print('Fold %d accuracy: %.2f' % (fold+1, accuracy_scores[-1]))
    print('Fold %d f1_score: %.2f' % (fold + 1, f1_scores[-1]))
# Print the mean accuracy score across all folds
print('Mean accuracy: %.2f' % np.mean(accuracy_scores))
print('Mean f1_score: %.2f' % np.mean(f1_scores))




Fold 1 accuracy: 0.97
Fold 1 f1_score: 0.97
Fold 2 accuracy: 0.97
Fold 2 f1_score: 0.97
Fold 3 accuracy: 0.97
Fold 3 f1_score: 0.97
Fold 4 accuracy: 0.97
Fold 4 f1_score: 0.97
Fold 5 accuracy: 0.97
Fold 5 f1_score: 0.97
Mean accuracy: 0.97
Mean f1_score: 0.97


cropped
Fold 1 f1_score: 0.97
Fold 2 accuracy: 0.97
Fold 2 f1_score: 0.97
Fold 3 accuracy: 0.97
Fold 3 f1_score: 0.97
Fold 4 accuracy: 0.97
Fold 4 f1_score: 0.97
Fold 5 accuracy: 0.97
Fold 5 f1_score: 0.97
Mean accuracy: 0.97
Mean f1_score: 0.97

Fold 1 accuracy: 0.96
Fold 1 f1_score: 0.96
Fold 2 accuracy: 0.96
Fold 2 f1_score: 0.96
Fold 3 accuracy: 0.96
Fold 3 f1_score: 0.96
Fold 4 accuracy: 0.96
Fold 4 f1_score: 0.96
Fold 5 accuracy: 0.96
Fold 5 f1_score: 0.96
Mean accuracy: 0.96
Mean f1_score: 0.96
